src/masterless.py

```diff
@@ -46,14 +56,50 @@ def ssh(stackname, node=None):
 #
 
 def parse_repo(string):
-    repo, rev = string.split('@')
-    return {
-        'repo': repo,
-        'rev': rev
-    }
+    if '@' not in string:
+        print 'skipping', string
+        return
+    return string.split('@')
 
 @task
 @echo_output
 def set_versions(stackname, *repolist):
-    bits = map(parse_repo, repolist)
-    return bits
+    "call with formula name and a revision, like: builder-private@ab87af78asdf2321431f31"
+    ctx = context_handler.load_context(stackname)
+
+    pdata = ctx['project']
+    known_formulas = pdata.get('formula-dependencies', [])
+    known_formulas = [
+        pdata['formula-repo'],
+        pdata['private-repo']
+    ]
+
+    known_formula_map = OrderedDict(zip(map(os.path.basename, known_formulas), known_formulas))
+
+    if not repolist:
+        return 'nothing to do'
+
+    arglist = []
+    for user_string in repolist:
+        if '@' not in user_string:
+            print 'skipping %r, no revision component' % user_string
+            continue
+
+        repo, rev = user_string.split('@')
+
+        if not rev.strip():
+            print 'skipping %r, empty revision component' % user_string
+            continue
+
+        if repo not in known_formula_map:
+            print 'skipping %r, unknown formula. known formulas: %s' % (repo, ', '.join(known_formula_map.keys()))
+            continue
+
+        arglist.append((repo, known_formula_map[repo], rev))
+
+    if not arglist:
+        return 'nothing to do'
+
+    with core.stack_conn(stackname):
```

Comment:
I think `stack_conn` doesn't cater for multiple nodes and `stack_all_ec2_nodes` should be used instead